stdlib/dynarray.mli

```diff
@@ -0,0 +1,418 @@
+
+val set : 'a t -> int -> 'a -> unit
+(** [set a i x] sets the [i]-th element of [a] to be [x].
+
+    Just like {!get}, [i] must be between [0] and [length a - 1]
+    included. [set] does not add new elements to the array -- see
+    {!add_last} to add an element.
+
+    @raise Invalid_argument if the index is invalid. *)
+
+val length : _ t -> int
+(** [length a] is the number of elements in the array.
+    The last element of [a], if not empty, is [get a (length a - 1)].
+    This operation is constant time. *)
+
+val is_empty : 'a t -> bool
+(** [is_empty a] is [true] if [a] is empty, that is, if [length a = 0]. *)
+
+val copy : 'a t -> 'a t
+(** [copy a] is a shallow copy of [a], that can be modified independently. *)
+
+
+(** {1:adding Adding elements}
+
+    Note: all operations adding elements can raise [Failure] if the
+    length would need to grow beyond {!Sys.max_array_length}. *)
+
+val add_last : 'a t -> 'a -> unit
+(** [add_last a x] adds the element [x] at the end of the array [a].
+    The length of [a] increases by [1]. *)
+
+val append_array : 'a t -> 'a array -> unit
+(** [append_array a b] adds all elements of [b] at the end of [a],
+    in the order they appear in [b].
+
+    For example, [a] will contain [1,2,3,4,5,6] after this code runs:
+    {[
+      let a = of_list [1;2;3];;
+      let () = append a [|4; 5; 6|];;
+    ]}
+*)
+
+val append_list : 'a t -> 'a list -> unit
+(** Like {!append_array} but with a list. *)
+
+val append : 'a t -> 'a t -> unit
+(** [append a b] is like [append_array a b],
+    but [b] is itself a dynamic arreay instead of a fixed-size array.
+
+    Beware! Calling [append a a] iterates on [a] and adds elements to
+    it at the same time; it is a programming error and fails with
+    [Invalid_argument].
+*)
+
+val append_seq : 'a t -> 'a Seq.t -> unit
+(** Like {!append_array} but with a sequence.
+
+    Beware! Calling [append_seq a (to_seq a)] is unspecified and may
+    result in an infinite loop, see the {!append} comment above.
+*)
+
+val append_iter :
+  'a t ->
+  (('a -> unit) -> 'iter-> unit) ->
+  'iter -> unit
+(** [append_iter a iter x] adds to [a] each element in [x]. It uses [iter]
+    to iterate over [x].
+
+    For example, [append_iter a List.iter [1;2;3]] would add elements
+    [1], [2], and [3] at the end of [a].
+    [append_iter a Queue.iter q] adds elements from the queue [q]. *)
+
+
+(** {1:removing Removing elements} *)
+
+val pop_last_opt : 'a t -> 'a option
+(** [pop_last_opt a] removes and returns the last element of [a],
+    or [None] if the array is empty. *)
+
+val pop_last : 'a t -> 'a
+(** [pop_last a] removes and returns the last element of [a].
+
+    @raise Not_found on an empty array. *)
+
+val remove_last : 'a t -> unit
+(** [remove_last a] removes the last element of [a] , or does nothing
+    if [is_empty a].
+*)
+
+val clear : 'a t -> unit
+(** [clear a] removes all the elements of [a].
+
+    It is equivalent to [truncate a 0].
```

Comment:
both are `()`, it's not very helpful to bring equational reasoning into the mix. I'd expect `clear a` to be slightly faster than `truncate a 0` because it doesn't have to handle the general case (it should be always inlined into a couple of instructions)

Reply:
I don't see the connection with `()`. It's always useful to understand functions in terms of others to reveal the primitives and their entanglements. 

If what you say is true of the implementation then this should rather be: 

```
(** [clear a] is [truncate a 0], but more efficient. *)
```

Reply:
Due to the "no leaks" guarantee, both `truncate` and `clear` are linear in the number of removed elements, it calls `Array.fill` to replace them all with empty slots. I don't think there is any performance difference between the two with the current implementation.